### PR TITLE
ensure consequent entries in consensus wal

### PR DESCRIPTION
**Pull request by generall**
### All Submissions:

- Raft paper proposes, that in case of leader conflict, entries with higher term should override submitted, but not yet committed entries in the consensus WAL.
- Previous implementation of consensus WAL didn't consider this case and only performed append to WAL, which caused inconsistency and undefined behavior of the raft-rs.
- This PR introduces index checks in the consensus WAL, allowing it to override existing entries based on raft index of the newly received entries.
- New code also assumes, the WAL index is not identical, but monotonous and does not contain any skips comparing to Raft index. In other words, WAL Index have a constant offset comparing to Raft index

Additional context:

reproduction of original bug involves multiple parallel consensus-related requests in combination with either high load, low memory, or very strict limitation of system resources (in our experiments, limiting to 5% of a single CPU). In this case the instances was slow enough to timeout on consensus heartbeat.

### CODE REVIEW

File: lib/storage/src/content_manager/consensus/consensus_wal.rs [hunk 1]
```diff
@@ -74,13 +78,171 @@ impl ConsensusOpWal {
         Ok(entry.transpose()?)
     }
 
+    /// Difference between raft index and WAL record number.
+    /// Difference might be different because of consensus snapshot.
+    fn index_offset(&self) -> Result<Option<u64>, StorageError> {
```
Comment (e-ivkov) [inline lib/storage/src/content_manager/consensus/consensus_wal.rs:83]: Can we use it also in `fn entry(&self, id: u64)`? There we also calculate the offset.

File: lib/storage/src/content_manager/consensus/consensus_wal.rs [hunk 2]
```diff
@@ -74,13 +78,171 @@ impl ConsensusOpWal {
...
+                    "Appending operation: term: {term}, index: {index} {operation:?}, entry: "
+                );
+            } else {
+                log::debug!("Appending entry: {entry:?}");
+            }
+
             let mut buf = vec![];
             entry.encode(&mut buf)?;
-            self.0.append(&buf)?;
+            let _wal_index = self.0.append(&buf)?;
```
Comment (e-ivkov) [inline lib/storage/src/content_manager/consensus/consensus_wal.rs:154]: Let's do a `debug_assert` here?

File: lib/storage/src/content_manager/consensus/consensus_wal.rs (hunk 1 above)
Comment (generall) [reply lib/storage/src/content_manager/consensus/consensus_wal.rs:83]: unfortunately no, it performs some extra intermediate operations

File: lib/storage/src/content_manager/consensus/consensus_wal.rs (hunk 2 above)
Comment (generall) [reply lib/storage/src/content_manager/consensus/consensus_wal.rs:151]: it's a non-trivial check, it depends on if there is a current offset

Comment (e-ivkov) [reply lib/storage/src/content_manager/consensus/consensus_wal.rs:151]: Yeah, I understand. We can do something like that:
```rust
#[cfg(debug_assertions)]
if ... {
   debug_assert!(...)
} else {
   debug_assert!(...)
}

Comment (generall) [reply lib/storage/src/content_manager/consensus/consensus_wal.rs]: :heavy_check_mark:

**Review by e-ivkov — APPROVED** [review]

**Pull request merged**